File: benchmark/adapters/evaluation.py
"""Public Prophet 1.4.0 evaluation mappings; no private diagnostics kernels or parallelism."""

from __future__ import annotations

import copy
import json
import math

import numpy as np
import pandas as pd
from prophet.diagnostics import cross_validation, performance_metrics


def metric_window(case):
    """Python's rolling window is fractional/weighted, unlike Effect whole horizon groups."""
    return {"overall": 1, "horizons": 0, "rolling": 0.5}[case["workload"]["metricAggregation"]]


def partition(dataset, case, records_frame):
    config = case["workload"]
    count = config.get("holdoutRows", 0)
    development = dataset["observations"][:-count] if count else dataset["observations"]
    holdout = dataset["observations"][-count:] if count else []
    return records_frame(development, include_value=True), records_frame(holdout, include_value=True)


def plan(case):
    config = case["workload"]["plan"]
    cutoffs = [pd.Timestamp(value).tz_localize(None) for value in config["cutoffs"]]
    horizon = pd.Timedelta(milliseconds=config["horizonMs"])
    return cutoffs, horizon


def cv(case, dataset, configure_model, fit_model, records_frame, intervals=False):
    history, _ = partition(dataset, case, records_frame)
    config = case["workload"]
    model = fit_model(history, case)  # Python diagnostics needs a fitted, mutable full-history model.
    if intervals:
        controls = config["interval"]
        model.uncertainty_samples = controls["samples"]
        model.interval_width = controls["intervalWidth"]
        np.random.seed(controls["seed"])
    cutoffs, horizon = plan(case)
    return cross_validation(model, horizon=horizon, cutoffs=cutoffs, parallel=None, disable_tqdm=True)


def search(case, dataset, configure_model, fit_model, records_frame):
    config = case["workload"]
    candidates = []
    for candidate in config["candidates"]:
        request = copy.deepcopy(case)
        if config["configuration"]["growth"] != "flat":
            request["workload"]["configuration"]["changepointPriorScale"] = candidate["changepointPriorScale"]
        if candidate.get("growth") == "logistic":
            request["workload"]["configuration"]["growth"] = "logistic"
        try:
            result = cv(request, dataset, configure_model, fit_model, records_frame)
            score = float(performance_metrics(result, metrics=["mae"], rolling_window=1)["mae"].iloc[0])
            if not math.isfinite(score):
                raise ValueError("Nonfinite search score")
            candidates.append({"id": candidate["id"], "score": score})
        except ValueError:
            if candidate.get("growth") != "logistic":
                raise
            candidates.append({"id": candidate["id"], "failure": "missing-capacity"})
    eligible = [index for index, item in enumerate(candidates) if "score" in item]
    winner = min(eligible, key=lambda index: (candidates[index]["score"], index))
    return candidates, winner


def holdout(case, dataset, configure_model, fit_model, records_frame, selection=None):
    history, assessment = partition(dataset, case, records_frame)
    candidates, winner = selection if selection is not None else search(case, dataset, configure_model, fit_model, records_frame)
    request = copy.deepcopy(case)
    if request["workload"]["configuration"]["growth"] != "flat":
        request["workload"]["configuration"]["changepointPriorScale"] = request["workload"]["candidates"][winner]["changepointPriorScale"]
    fitted = fit_model(history, request)
    predicted = fitted.predict(assessment.drop(columns="y"))
    actual = assessment["y"].to_numpy()
    forecast = predicted["yhat"].to_numpy()
    baselines = {"last-observation": float(history["y"].iloc[-1]), "training-mean": float(history["y"].mean())}
    return {"selected": candidates[winner], "candidates": candidates,
            "actual": actual.tolist(), "predicted": forecast.tolist(),
            "mae": float(np.mean(np.abs(actual - forecast))), "baselines": baselines}


def operation(case, dataset, phase, configure_model, fit_model, records_frame):
    history, _ = partition(dataset, case, records_frame)
    if phase == "evaluation-input-conversion":
        return history
    cutoffs, horizon = plan(case)
    if phase == "evaluation-plan":
        return [{"cutoff": cutoff.isoformat(), "training": int((history["ds"] <= cutoff).sum()),
                 "assessment": int(((history["ds"] > cutoff) & (history["ds"] <= cutoff + horizon)).sum())}
                for cutoff in cutoffs]
    if phase in ("evaluation-point", "cold-first-evaluation"):
        return cv(case, dataset, configure_model, fit_model, records_frame)
    if phase == "evaluation-intervals":
        return cv(case, dataset, configure_model, fit_model, records_frame, intervals=True)
    if phase == "evaluation-metrics":
        result = cv(case, dataset, configure_model, fit_model, records_frame)
        return performance_metrics(result, metrics=["mae", "rmse", "smape", "mape"], rolling_window=metric_window(case))
    if phase == "evaluation-baseline":
        result = []
        for cutoff in cutoffs:
            training = history.loc[history["ds"] <= cutoff]
            assessment = history.loc[(history["ds"] > cutoff) & (history["ds"] <= cutoff + horizon)]
            values = dict(zip(training["ds"], training["y"]))
            seasonal = [float(values[timestamp - pd.Timedelta(days=7)]) for timestamp in assessment["ds"]]
            result.append({"cutoff": cutoff.isoformat(), "last": [float(training["y"].iloc[-1])] * len(assessment),
                           "mean": [float(training["y"].mean())] * len(assessment), "seasonal": seasonal})
        return result
    if phase == "evaluation-search":
        return search(case, dataset, configure_model, fit_model, records_frame)
    if phase == "evaluation-holdout":
        return holdout(case, dataset, configure_model, fit_model, records_frame)
    if phase == "evaluation-report-encode":
        return json.dumps(holdout(case, dataset, configure_model, fit_model, records_frame), allow_nan=False)
    if phase == "evaluation-report-decode":
        return json.loads(json.dumps(holdout(case, dataset, configure_model, fit_model, records_frame), allow_nan=False))
    raise ValueError(f"Unsupported evaluation phase: {phase}")


def prepare_operation(case, dataset, phase, configure_model, fit_model, records_frame):
    """Exclude CV/search setup from independent pure-phase boundaries."""
    if phase == "evaluation-metrics":
        result = cv(case, dataset, configure_model, fit_model, records_frame)
        return lambda: performance_metrics(result, metrics=["mae", "rmse", "smape", "mape"], rolling_window=metric_window(case))
    if phase == "evaluation-holdout":
        selection = search(case, dataset, configure_model, fit_model, records_frame)
        return lambda: holdout(case, dataset, configure_model, fit_model, records_frame, selection)
    if phase == "evaluation-report-encode":
        report = holdout(case, dataset, configure_model, fit_model, records_frame)
        return lambda: json.dumps(report, allow_nan=False)
    if phase == "evaluation-report-decode":
        encoded = json.dumps(holdout(case, dataset, configure_model, fit_model, records_frame), allow_nan=False)
        return lambda: json.loads(encoded)
    return lambda: operation(case, dataset, phase, configure_model, fit_model, records_frame)


def correctness(case, dataset, configure_model, fit_model, records_frame, run):
    history, _ = partition(dataset, case, records_frame)
    cutoffs, horizon = plan(case)
    result = cv(case, dataset, configure_model, fit_model, records_frame)
    rows = []
    training = 0
    assessment = 0
    for cutoff in cutoffs:
        expected = history.loc[(history["ds"] > cutoff) & (history["ds"] <= cutoff + horizon)]
        training += int((history["ds"] <= cutoff).sum())
        assessment += len(expected)
        observed = result.loc[result["cutoff"] == cutoff]
        if list(observed["ds"]) != list(expected["ds"]) or list(observed["y"]) != list(expected["y"]):
            raise ValueError("Python CV cutoff, order or target alignment differs")
        for row in observed.itertuples():
            if not math.isfinite(row.yhat):
                raise ValueError("Python CV forecast is nonfinite")
            rows.append({"cutoff": int(cutoff.value // 1_000_000), "timestamp": int(row.ds.value // 1_000_000),
                         "actual": float(row.y), "predicted": float(row.yhat)})
    if len(rows) != assessment:
        raise ValueError("Python CV assessment dimensions differ")
    if "evaluation-intervals" in case["phases"]:
        interval = operation(case, dataset, "evaluation-intervals", configure_model, fit_model, records_frame)
        if len(interval) != assessment or not np.isfinite(interval[["yhat_lower", "yhat_upper"]].to_numpy()).all() or (interval["yhat_lower"] > interval["yhat_upper"]).any():
            raise ValueError("Python intervals failed finite dimension/order checks")
    for phase in case["phases"]:
        if phase in ("evaluation-search", "evaluation-holdout", "evaluation-report-encode", "evaluation-report-decode"):
            outcome = operation(case, dataset, phase, configure_model, fit_model, records_frame)
            if phase == "evaluation-search" and len([candidate for candidate in outcome[0] if "failure" in candidate]) != case["workload"].get("expectedFailures", 0):
                raise ValueError("Unexpected Python candidate failure count")
    mae = float(np.mean([abs(row["actual"] - row["predicted"]) for row in rows]))
    return {"caseId": case["id"], "run": run, "status": "locally-passed",
            "modelKind": case["workload"]["configuration"]["growth"],
            "changepointTimestamps": [], "seasonalities": [], "events": [], "regressors": [], "forecasts": [],
            "evaluation": {"cutoffs": [int(cutoff.value // 1_000_000) for cutoff in cutoffs],
                           "trainingRows": training, "assessmentRows": assessment,
                           "rows": rows, "mae": mae}}

File: benchmark/evaluation.ts
import { Effect, Layer } from "effect";
import {
  crossValidate,
  crossValidateBaseline,
  decodeEvaluationReport,
  encodeEvaluationReport,
  evaluateHoldout,
  performanceMetrics,
  planRollingOrigin,
  prophetFittingBackendLayer,
  searchModels,
  type EncodedProphetOptions,
  type ModelSearchResult,
} from "effect-prophet";

import type { BenchmarkCase, BenchmarkDataset, BenchmarkPhase } from "./case.ts";
import { effectOptionsForCase } from "./effect-case.ts";
import type { CorrectnessProjection } from "./result.ts";

const run = <A, E>(effect: Effect.Effect<A, E>): A =>
  Effect.runSync(effect.pipe(Effect.withTracerEnabled(false)));

const runFitting = <A, E>(
  effect: Effect.Effect<A, E, Layer.Success<typeof prophetFittingBackendLayer>>,
): A => run(effect.pipe(Effect.provide(prophetFittingBackendLayer)));

// Holdout is removed before search, planning and every fold fit.
const inputs = (benchmarkCase: BenchmarkCase, dataset: BenchmarkDataset) => {
  const workload = benchmarkCase.workload;

  if (workload.kind !== "evaluation") {
    throw new Error("Expected an evaluation workload");
  }

  const development = dataset.observations.slice(
    0,
    dataset.observations.length - (workload.holdoutRows ?? 0),
  );

  const holdout = dataset.observations.slice(development.length);
  const options = effectOptionsForCase(benchmarkCase);

  if (options === undefined) {
    throw new Error("Expected MAP options");
  }

  return {
    workload,
    development,
    holdout,
    options,
    plan: {
      horizonMs: workload.plan.horizonMs,
      cutoffs: { mode: "explicit" as const, timestamps: workload.plan.cutoffs },
    },
  };
};

const metricPolicy = (workload: Extract<BenchmarkCase["workload"], { kind: "evaluation" }>) => ({
  metrics: ["mae", "rmse", "smape", "mape"] as const,
  aggregation:
    workload.metricAggregation === "rolling"
      ? { kind: "rolling" as const, window: { kind: "fraction" as const, value: 0.5 } }
      : { kind: workload.metricAggregation },
  mapeZeroActual: workload.mapeZeroActual,
});

const candidateOptions = (
  options: EncodedProphetOptions,
  prior: number,
  growth?: "logistic",
): EncodedProphetOptions => {
  if (options.growth === "flat") return options;

  if (growth === "logistic" && options.growth === "linear") {
    return {
      ...options,
      growth: "logistic",
      map: { ...options.map, changepointPriorScale: prior },
    };
  }

  return { ...options, map: { ...options.map, changepointPriorScale: prior } };
};

const search = (benchmarkCase: BenchmarkCase, dataset: BenchmarkDataset): ModelSearchResult => {
  const { workload, development, options, plan } = inputs(benchmarkCase, dataset);

  return runFitting(
    searchModels(development, {
      candidates: (workload.candidates ?? []).map((candidate) => ({
        id: candidate.id,
        options: candidateOptions(options, candidate.changepointPriorScale, candidate.growth),
      })),
      plan,
      objective: { metric: "mae", aggregation: { kind: "overall" }, direction: "minimize" },
      failurePolicy: "record",
    }),
  );
};

const holdoutUsingSearch = (
  benchmarkCase: BenchmarkCase,
  dataset: BenchmarkDataset,
  result: ModelSearchResult,
) => {
  const { development, holdout: assessment } = inputs(benchmarkCase, dataset);
  const selected = result.candidates[result.selected.candidateIndex];
  const firstDevelopment = development[0];
  const firstHoldout = assessment[0];

  if (firstDevelopment === undefined || firstHoldout === undefined) {
    throw new Error("Holdout benchmark requires both partitions");
  }

  if (selected?.kind !== "success") throw new Error("Search did not select a successful candidate");

  return runFitting(
    evaluateHoldout({
      development: [firstDevelopment, ...development.slice(1)],
      holdout: [firstHoldout, ...assessment.slice(1)],
      search: result,
      selectedCandidate: {
        id: selected.id,
        candidateIndex: selected.candidateIndex,
        options: selected.options,
        score: selected.score,
      },
      metrics: { metrics: ["mae"], aggregation: { kind: "overall" } },
      baselines: [{ kind: "last-observation" }, { kind: "training-mean" }],
      provenance: { build: "benchmark-release", environment: "benchmark-container" },
    }),
  );
};

const holdout = (benchmarkCase: BenchmarkCase, dataset: BenchmarkDataset) =>
  holdoutUsingSearch(benchmarkCase, dataset, search(benchmarkCase, dataset));

/** Untimed fold identity, alignment, metric and replay checks before eligibility. */
export const evaluationCorrectness = (
  benchmarkCase: BenchmarkCase,
  dataset: BenchmarkDataset,
  runIndex: number,
): CorrectnessProjection => {
  const { workload, development, options, plan } = inputs(benchmarkCase, dataset);
  const summary = run(planRollingOrigin(development, options, plan));
  const result = runFitting(crossValidate(development, options, plan));

  const rows = result.rows.map((row) => ({
    cutoff: row.cutoff,
    timestamp: row.timestamp,
    actual: row.actual,
    predicted: row.predicted,
  }));

  const expected = summary.folds.reduce((count, fold) => count + fold.assessmentCount, 0);

  if (result.rows.length !== expected || result.folds.length !== summary.folds.length) {
    throw new Error("Evaluation fold/row counts disagree with the public plan");
  }

  for (const [index, row] of result.rows.entries()) {
    const actual = development.find((item) => Date.parse(item.timestamp) === row.timestamp);

    if (
      actual?.value !== row.actual ||
      !Number.isFinite(row.predicted) ||
      row.horizonMs !== row.timestamp - row.cutoff ||
      (index > 0 && row.fold < (result.rows[index - 1]?.fold ?? 0))
    ) {
      throw new Error("Evaluation output failed alignment or finite checks");
    }
  }

  const metrics = run(
    performanceMetrics(result, { metrics: ["mae"], aggregation: { kind: "overall" } }),
  );

  if (metrics.kind !== "overall") throw new Error("Expected overall evaluation metrics");

  const mae = metrics.bucket.scores[0]?.value;

  if (mae === undefined || !Number.isFinite(mae)) throw new Error("Evaluation MAE is not finite");

  for (const phase of benchmarkCase.phases) {
    if (phase === "evaluation-intervals") {
      if (workload.interval === undefined) throw new Error("Missing interval controls");

      const interval = runFitting(
        crossValidate(development, options, plan, {
          mode: "intervals",
          uncertainty: workload.interval,
        }),
      );

      if (
        interval.rows.length !== result.rows.length ||
        interval.rows.some(
          (row, index) =>
            row.timestamp !== result.rows[index]?.timestamp ||
            row.cutoff !== result.rows[index]?.cutoff ||
            !Number.isFinite(row.lower) ||
            !Number.isFinite(row.upper) ||
            row.lower > row.upper,
        )
      ) {
        throw new Error("Interval dimensions or bounds differ from point CV");
      }
    } else if (
      phase === "evaluation-search" ||
      phase === "evaluation-holdout" ||
      phase === "evaluation-report-encode" ||
      phase === "evaluation-report-decode"
    ) {
      if (phase === "evaluation-search") {
        const selection = search(benchmarkCase, dataset);

        if (
          selection.candidates.filter((candidate) => candidate.kind === "failure").length !==
          (workload.expectedFailures ?? 0)
        ) {
          throw new Error("Search recorded an unexpected candidate failure count");
        }
      } else {
        evaluationOperation(benchmarkCase, dataset, phase);
      }
    }
  }

  return {
    caseId: benchmarkCase.id,
    run: runIndex,
    status: "locally-passed",
    modelKind: workload.configuration.growth,
    changepointTimestamps: [],
    seasonalities: [],
    events: [],
    regressors: [],
    forecasts: [],
    evaluation: {
      cutoffs: Array.from(summary.cutoffs),
      trainingRows: summary.folds.reduce((count, fold) => count + fold.trainingCount, 0),
      assessmentRows: expected,
      rows,
      mae,
    },
  };
};

/** Prepare untimed inputs for pure phases so fitting is excluded from their samples. */
export const preparedEvaluationOperation = (
  benchmarkCase: BenchmarkCase,
  dataset: BenchmarkDataset,
  phase: BenchmarkPhase,
): (() => string | ReturnType<typeof evaluationOperation>) => {
  const { workload, development, options, plan } = inputs(benchmarkCase, dataset);

  if (phase === "evaluation-metrics") {
    const result = runFitting(crossValidate(development, options, plan));

    return () => run(performanceMetrics(result, metricPolicy(workload)));
  }

  if (phase === "evaluation-baseline") {
    const summary = run(planRollingOrigin(development, options, plan));

    return () => [
      run(crossValidateBaseline(development, summary, { kind: "last-observation" })),
      run(crossValidateBaseline(development, summary, { kind: "training-mean" })),
      run(
        crossValidateBaseline(development, summary, {
          kind: "seasonal-naive",
          lagMs: 7 * 86_400_000,
        }),
      ),
    ];
  }

  if (phase === "evaluation-holdout") {
    const selection = search(benchmarkCase, dataset);

    return () => holdoutUsingSearch(benchmarkCase, dataset, selection);
  }

  if (phase === "evaluation-report-encode") {
    const report = holdout(benchmarkCase, dataset);

    return () => JSON.stringify(run(encodeEvaluationReport(report)));
  }

  if (phase === "evaluation-report-decode") {
    const report = holdout(benchmarkCase, dataset);
    const encoded = JSON.stringify(run(encodeEvaluationReport(report)));

    return () => run(decodeEvaluationReport(JSON.parse(encoded)));
  }

  return () => evaluationOperation(benchmarkCase, dataset, phase);
};

/** Execute one complete, tracing-disabled public evaluation operation. */
export const evaluationOperation = (
  benchmarkCase: BenchmarkCase,
  dataset: BenchmarkDataset,
  phase: BenchmarkPhase,
) => {
  const { workload, development, options, plan } = inputs(benchmarkCase, dataset);

  switch (phase) {
    case "evaluation-plan":
      return run(planRollingOrigin(development, options, plan));
    case "evaluation-point":
    case "cold-first-evaluation":
      return runFitting(crossValidate(development, options, plan));
    case "evaluation-intervals":
      if (workload.interval === undefined) throw new Error("Missing interval controls");

      return runFitting(
        crossValidate(development, options, plan, {
          mode: "intervals",
          uncertainty: workload.interval,
        }),
      );
    case "evaluation-metrics": {
      const result = runFitting(crossValidate(development, options, plan));

      return run(performanceMetrics(result, metricPolicy(workload)));
    }

    case "evaluation-baseline": {
      const summary = run(planRollingOrigin(development, options, plan));

      return [
        run(crossValidateBaseline(development, summary, { kind: "last-observation" })),
        run(crossValidateBaseline(development, summary, { kind: "training-mean" })),
        run(
          crossValidateBaseline(development, summary, {
            kind: "seasonal-naive",
            lagMs: 7 * 86_400_000,
          }),
        ),
      ];
    }

    case "evaluation-search":
      return search(benchmarkCase, dataset);
    case "evaluation-holdout":
      return holdout(benchmarkCase, dataset);
    case "evaluation-report-encode":
    case "evaluation-report-decode": {
      const report = holdout(benchmarkCase, dataset);
      const encoded = run(encodeEvaluationReport(report));

      return run(decodeEvaluationReport(JSON.parse(JSON.stringify(encoded))));
    }

    default:
      throw new Error(`Not an evaluation phase: ${phase}`);
  }
};
